### PR TITLE
Shard-specific TXGen

### PR DESCRIPTION
**Pull request by alajko**
## Issue

Generate a TXGenerator for specific shards

#### Test Coverage Data

Screenshots attached below.

## Future Work
- [ ] Remove `consensusMessageHandler` timeout message
- [ ] Increase Txn Throughput. Currently high tx throughput might be knocking down network -- need to investigate
- [ ] Improve per second throughput.
- [ ] Investigate why sending to shard 0, sends txns to all shards.

### CODE REVIEW

**Comment by alajko** [comment]
shard-specific-txgen-2

**Comment by alajko** [comment]
shard-specific-txgen-2

**Comment by alajko** [comment]
Tested
- [ ] Process quits after duration
- [ ] TXGen works on shard 0

**Comment by alajko** [comment]
shard0 txgen also works:

**Comment by alajko** [comment]
shard0-sends-txns-to-everyone

File: api/service/syncing/syncing.go [hunk 1]
```diff
@@ -607,6 +608,13 @@ func (ss *StateSync) getMaxPeerHeight() uint64 {
 	return maxHeight
 }
 
+// IsItSameBlockchainHeight checks whether the node is out of sync from other peers
+func (ss *StateSync) IsItSameBlockchainHeight(bc *core.BlockChain) (uint64, bool) {
```
Comment (rlan35) [inline api/service/syncing/syncing.go]: Change to "IsSameBlockchainHeight"

File: cmd/client/txgen/README.md [hunk 2]
```diff
@@ -1 +1,2 @@
-The txgen program is used to simulate transactions and hit the Harmony network to loadtest its performance and robustness.
\ No newline at end of file
+The txgen program is used to simulate transactions and hit the Harmony network to loadtest its performance and robustness.
+You can send txns to specific shards 1,2,3 or to shard 0. Sending it to shard 0, broadcasts txns to all the shards.
```
Comment (rlan35) [inline cmd/client/txgen/README.md]: broadcasts txns to all shards is a unwanted effect we should fix. Please add a note or todo.

File: cmd/client/txgen/main.go [hunk 3]
```diff
@@ -60,14 +60,14 @@ func printVersion(me string) {
...
-	port              = flag.String("port", "9999", "port of the node.")
-	maxNumTxsPerBatch = flag.Int("max_num_txs_per_batch", 20000, "number of transactions to send per message")
-	logFolder         = flag.String("log_folder", "latest", "the folder collecting the logs of this execution")
-	duration          = flag.Int("duration", 30, "duration of the tx generation in second. If it's negative, the experiment runs forever.")
-	versionFlag       = flag.Bool("version", false, "Output version info")
-	crossShardRatio   = flag.Int("cross_shard_ratio", 30, "The percentage of cross shard transactions.") //Keeping this for backward compatibility
-	shardIDFlag       = flag.Int("shardID", 0, "The shardID the node belongs to.")
+	ip              = flag.String("ip", "127.0.0.1", "IP of the node")
+	port            = flag.String("port", "9999", "port of the node.")
+	numTxns         = flag.Int("numTxns", 100, "number of transactions to send per message should be divisble by 100")
```
Comment (rlan35) [inline cmd/client/txgen/main.go]: why divisible by 100?

File: cmd/client/txgen/main.go [hunk 4]
```diff
@@ -232,24 +251,33 @@ pushLoop:
...
 	}
 }
 
 // GenerateSimulatedTransactionsAccount generates simulated transaction for account model.
 func GenerateSimulatedTransactionsAccount(shardID uint32, node *node.Node, setting Settings) (types.Transactions, error) {
-	_ = setting // TODO: make use of settings
-	txs := make([]*types.Transaction, 100)
+	TxnsToGenerate := setting.MaxNumTxsPerBatch // TODO: make use of settings
+	txs := make([]*types.Transaction, TxnsToGenerate)
+	rounds := (TxnsToGenerate / 100)
```
Comment (rlan35) [inline cmd/client/txgen/main.go]: I see the reason for visible by 100. But it shouldn't be a hard requirement. It should be able to send any amount. Also why multiple rounds?

File: cmd/client/txgen/README.md (hunk 2 above)
Comment (alajko) [reply cmd/client/txgen/README.md]: https://github.com/harmony-one/harmony/issues/847

Comment (alajko) [reply cmd/client/txgen/README.md]: Done.

File: cmd/client/txgen/main.go (hunk 4 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: I copied over what we had earlier. With rounds we calculate `basenonce` once and then add the `rounds` number to adjust the correct nonce.

File: api/service/syncing/syncing.go (hunk 1 above)
Comment (alajko) [reply api/service/syncing/syncing.go]: Done.

File: cmd/client/txgen/main.go (hunk 3 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: This is no longer required.

File: cmd/client/txgen/main.go (hunk 4 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: This code is now changed.

**Comment by alajko** [comment]
Jenkins, test this please

**Pull request merged**